File: 1376-time-needed-to-inform-all-employees/1376-time-needed-to-inform-all-employees.py
from collections import deque

class Solution:
    def numOfMinutes(self, n: int, headID: int, manager: List[int], informTime: List[int]) -> int:
        graph = [[] for _ in range(n)]
        
        for i, head in enumerate(manager):
            if head == -1:
                continue
            graph[head].append(i)
            
        
        q = deque([(headID, informTime[headID])])
        answer = 0
        
        while q:
            now, time = q.popleft()
            answer = max(answer, time)
            for _next in graph[now]:
                q.append((_next, time + informTime[_next]))
            
        
        return answer

File: 1376-time-needed-to-inform-all-employees/1376-time-needed-to-inform-all-employees.rs
use std::{collections::VecDeque, cmp};

impl Solution {
    pub fn num_of_minutes(n: i32, head_id: i32, manager: Vec<i32>, inform_time: Vec<i32>) -> i32 {
        let n = n as usize;
        let mut graph: Vec<Vec<usize>> = vec![vec![]; n];
        
        for (i, num) in manager.into_iter().enumerate() {
            if num == -1 {
                continue;
            }
            graph[num as usize].push(i as usize);
        }
        
        let mut q: VecDeque<(usize, i32)> = VecDeque::new();
        q.push_back((head_id as usize , inform_time[head_id as usize]));
        
        let mut answer = 0;
        
        while let Some((now, time)) = q.pop_front() {
            answer = cmp::max(answer, time);
            
            for next in graph[now as usize].iter() {
                q.push_back((*next, time + inform_time[*next]));
            }
        }
        return answer;
    }
}